### DISCUSSION TITLE
Model compostion and Optional Properties

### DISCUSSION THREAD

**Discussion by laduke** [microsoft/typespec#2814] [Q&A] [answered]
Dealing with an existing API. 

There's a model like below. Then in HTTP stuff , the POST request format is ThingWritable. The response format is Thing and the GET response is Thing.
The little issue is Thing has ThingWritable's props as Optional. They should be required there. 

What's the good pattern for composing something like this?

```
model ThingWritable {
  prop1?: boolean;
  prop2?: boolean;
}

model Thing{
  ...ThingWritable;
  readonly1: boolean;
  readonly2: boolean;
}
```

**Comment by bterlson** [comment, marked as answer]
We have [`OptionalProperties<T>`](https://typespec.io/docs/standard-library/built-in-data-types#OptionalProperties), though not `RequiredProperties` which should be added I think, so you could write:

```
model Thing {
    ... RequiredProperties<ThingWritable>;
}
```

For now, you could do something like:

```
alias ThingProps = {
  prop1: boolean;
  prop2: boolean;
}

model ThingWritable {
   ... OptionalProperties<ThingProps>;
}

model Thing {
  ... ThingProps;
}
```

**Comment by bterlson** [comment]
I should also say, ideally you use [visibility](https://typespec.io/docs/standard-library/built-in-decorators#@visibility) for this purpose but that can be a challenge when adapting existing APIs.

**Comment by laduke** [comment]
Perfect. Thank you for both of those tips.
